### ISSUE TITLE
Why swapping nn.Linear with FP8Linear brings throughput decrease for 7B LLAMA2-like model?

### ISSUE THREAD

**Issue by zigzagcai** [pytorch/ao#1199]
Hi developers,

Thanks for such a great project that provides elegant interface for FP8 training! 

For simplicity, I use dynamic scaling with FP8Linear. But when I tried to integrate FP8 functionalities with our training framework, I found the throughput of linear modules (such like the feedforward) decrease a lot.

- traces that use naive bfloat16 `torch.nn.Linear`

  - CPU
![image](https://github.com/user-attachments/assets/412e7975-87a9-45a8-a696-86c760b25699)
  - GPU
![image](https://github.com/user-attachments/assets/60109bce-dfcd-48d6-a50e-219d335d8c34)

- traces that swap `torch.nn.Linear` with `FP8Linear`, via [convert_to_float8_training ](https://github.com/pytorch/ao/blob/main/torchao/float8/float8_linear_utils.py#L125)API

  - CPU
![image](https://github.com/user-attachments/assets/8233da85-6d1f-4afb-9262-b8b7656b661b)

- Code to reproduce the performance results: https://github.com/InternLM/InternEvo/tree/fp8_enable

- Test environment: 
  - 1x node, with 2x H100 GPUs
  - torch 2.5.0
  - train script: `torchrun  --nproc_per_node=2 --nnodes=1 --node_rank=0  train.py --config configs/7B_sft.py --launcher torch --seed 1024`

Thanks in advance if anyone could provide some insights!

**Comment by vkuzo** [comment]
hi @zigzagcai , from the trace above and from the code in https://github.com/InternLM/InternEvo/commit/e6f5562551e50adb0fd2f424a4d4ab4b9a7984a6#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95R22 I would guess that you are running in eager mode.  This is expected to be slow as the float8 ops are not fused with surrounding ops.  We require torch.compile to achieve good performance, have you tried enabling that?  From your commit, the change would be

```
    float8_handler = Float8Handler()
    float8_handler.convert_to_float8_training(model)
    # enable torch.compile for improved performance
    model = torch.compile(model)
```

You can also try enabling torch.compile on smaller parts of the model if full-model compile is problematic.  Here is an example of doing this from our torchtitan repository: https://github.com/pytorch/torchtitan/blob/dbb0520d51f3f92b5391b46b545959c7dcebe197/torchtitan/parallelisms/parallelize_llama.py#L304

**Comment by zigzagcai** [comment]
> hi @zigzagcai , from the trace above and from the code in [InternLM/InternEvo@e6f5562#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95R22](https://github.com/InternLM/InternEvo/commit/e6f5562551e50adb0fd2f424a4d4ab4b9a7984a6#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95R22) I would guess that you are running in eager mode. This is expected to be slow as the float8 ops are not fused with surrounding ops. We require torch.compile to achieve good performance, have you tried enabling that? From your commit, the change would be
> 
> ```
>     float8_handler = Float8Handler()
>     float8_handler.convert_to_float8_training(model)
>     # enable torch.compile for improved performance
>     model = torch.compile(model)
> ```
> 
> You can also try enabling torch.compile on smaller parts of the model if full-model compile is problematic. Here is an example of doing this from our torchtitan repository: https://github.com/pytorch/torchtitan/blob/dbb0520d51f3f92b5391b46b545959c7dcebe197/torchtitan/parallelisms/parallelize_llama.py#L304

Hi @vkuzo 

Thanks! `model = torch.compile(model)` do helps! The FP8Linear throughput improved and show performance gain above naive bfloat16 `nn.Linear`.

**Comment by vkuzo** [comment]
awesome! Feel free to let us know with more questions!
